// @triton-one/yellowstone-grpc/dist/cjs/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.txErrDecode = exports.txDeshredEncode = exports.txEncode = exports.ClientDeshredDuplexStream = exports.ClientDuplexStream = exports.SubscribeUpdateTransactionInfo = exports.SubscribeUpdateTransaction = exports.SubscribeUpdateSlot = exports.SubscribeUpdatePing = exports.SubscribeUpdateDeshredTransactionInfo = exports.SubscribeUpdateDeshredTransaction = exports.SubscribeUpdateDeshred = exports.SubscribeUpdateBlockMeta = exports.SubscribeUpdateBlock = exports.SubscribeUpdateAccountInfo = exports.SubscribeUpdateAccount = exports.SubscribeUpdate = exports.SubscribeRequestFilterTransactions = exports.SubscribeRequestFilterSlots = exports.SubscribeRequestFilterEntry = exports.SubscribeRequestFilterBlocksMeta = exports.SubscribeRequestFilterBlocks = exports.SubscribeRequestFilterAccountsFilterMemcmp = exports.SubscribeRequestFilterAccountsFilterLamports = exports.SubscribeRequestFilterAccountsFilter = exports.SubscribeRequestFilterAccounts = exports.SubscribeRequestAccountsDataSlice = exports.SubscribeRequest_TransactionsEntry = exports.SubscribeRequest_SlotsEntry = exports.SubscribeRequestFilterDeshredTransactions = exports.SubscribeRequest_BlocksMetaEntry = exports.SubscribeRequest_BlocksEntry = exports.SubscribeRequest_AccountsEntry = exports.SubscribeRequest = exports.SubscribeDeshredRequest_DeshredTransactionsEntry = exports.SubscribeDeshredRequest = exports.CommitmentLevel = void 0;
var geyser_1 = require("./grpc/geyser");
var geyser_2 = require("./grpc/geyser");
Object.defineProperty(exports, "CommitmentLevel", { enumerable: true, get: function () { return geyser_2.CommitmentLevel; } });
Object.defineProperty(exports, "SubscribeDeshredRequest", { enumerable: true, get: function () { return geyser_2.SubscribeDeshredRequest; } });
Object.defineProperty(exports, "SubscribeDeshredRequest_DeshredTransactionsEntry", { enumerable: true, get: function () { return geyser_2.SubscribeDeshredRequest_DeshredTransactionsEntry; } });
Object.defineProperty(exports, "SubscribeRequest", { enumerable: true, get: function () { return geyser_2.SubscribeRequest; } });
Object.defineProperty(exports, "SubscribeRequest_AccountsEntry", { enumerable: true, get: function () { return geyser_2.SubscribeRequest_AccountsEntry; } });
Object.defineProperty(exports, "SubscribeRequest_BlocksEntry", { enumerable: true, get: function () { return geyser_2.SubscribeRequest_BlocksEntry; } });
Object.defineProperty(exports, "SubscribeRequest_BlocksMetaEntry", { enumerable: true, get: function () { return geyser_2.SubscribeRequest_BlocksMetaEntry; } });
Object.defineProperty(exports, "SubscribeRequestFilterDeshredTransactions", { enumerable: true, get: function () { return geyser_2.SubscribeRequestFilterDeshredTransactions; } });
Object.defineProperty(exports, "SubscribeRequest_SlotsEntry", { enumerable: true, get: function () { return geyser_2.SubscribeRequest_SlotsEntry; } });
Object.defineProperty(exports, "SubscribeRequest_TransactionsEntry", { enumerable: true, get: function () { return geyser_2.SubscribeRequest_TransactionsEntry; } });
Object.defineProperty(exports, "SubscribeRequestAccountsDataSlice", { enumerable: true, get: function () { return geyser_2.SubscribeRequestAccountsDataSlice; } });
Object.defineProperty(exports, "SubscribeRequestFilterAccounts", { enumerable: true, get: function () { return geyser_2.SubscribeRequestFilterAccounts; } });
Object.defineProperty(exports, "SubscribeRequestFilterAccountsFilter", { enumerable: true, get: function () { return geyser_2.SubscribeRequestFilterAccountsFilter; } });
Object.defineProperty(exports, "SubscribeRequestFilterAccountsFilterLamports", { enumerable: true, get: function () { return geyser_2.SubscribeRequestFilterAccountsFilterLamports; } });
Object.defineProperty(exports, "SubscribeRequestFilterAccountsFilterMemcmp", { enumerable: true, get: function () { return geyser_2.SubscribeRequestFilterAccountsFilterMemcmp; } });
Object.defineProperty(exports, "SubscribeRequestFilterBlocks", { enumerable: true, get: function () { return geyser_2.SubscribeRequestFilterBlocks; } });
Object.defineProperty(exports, "SubscribeRequestFilterBlocksMeta", { enumerable: true, get: function () { return geyser_2.SubscribeRequestFilterBlocksMeta; } });
Object.defineProperty(exports, "SubscribeRequestFilterEntry", { enumerable: true, get: function () { return geyser_2.SubscribeRequestFilterEntry; } });
Object.defineProperty(exports, "SubscribeRequestFilterSlots", { enumerable: true, get: function () { return geyser_2.SubscribeRequestFilterSlots; } });
Object.defineProperty(exports, "SubscribeRequestFilterTransactions", { enumerable: true, get: function () { return geyser_2.SubscribeRequestFilterTransactions; } });
Object.defineProperty(exports, "SubscribeUpdate", { enumerable: true, get: function () { return geyser_2.SubscribeUpdate; } });
Object.defineProperty(exports, "SubscribeUpdateAccount", { enumerable: true, get: function () { return geyser_2.SubscribeUpdateAccount; } });
Object.defineProperty(exports, "SubscribeUpdateAccountInfo", { enumerable: true, get: function () { return geyser_2.SubscribeUpdateAccountInfo; } });
Object.defineProperty(exports, "SubscribeUpdateBlock", { enumerable: true, get: function () { return geyser_2.SubscribeUpdateBlock; } });
Object.defineProperty(exports, "SubscribeUpdateBlockMeta", { enumerable: true, get: function () { return geyser_2.SubscribeUpdateBlockMeta; } });
Object.defineProperty(exports, "SubscribeUpdateDeshred", { enumerable: true, get: function () { return geyser_2.SubscribeUpdateDeshred; } });
Object.defineProperty(exports, "SubscribeUpdateDeshredTransaction", { enumerable: true, get: function () { return geyser_2.SubscribeUpdateDeshredTransaction; } });
Object.defineProperty(exports, "SubscribeUpdateDeshredTransactionInfo", { enumerable: true, get: function () { return geyser_2.SubscribeUpdateDeshredTransactionInfo; } });
Object.defineProperty(exports, "SubscribeUpdatePing", { enumerable: true, get: function () { return geyser_2.SubscribeUpdatePing; } });
Object.defineProperty(exports, "SubscribeUpdateSlot", { enumerable: true, get: function () { return geyser_2.SubscribeUpdateSlot; } });
Object.defineProperty(exports, "SubscribeUpdateTransaction", { enumerable: true, get: function () { return geyser_2.SubscribeUpdateTransaction; } });
Object.defineProperty(exports, "SubscribeUpdateTransactionInfo", { enumerable: true, get: function () { return geyser_2.SubscribeUpdateTransactionInfo; } });
var stream_1 = require("stream");
var napi = __importStar(require("./napi/index"));
function fromJsSubscribeUpdate(update) {
    var _a, _b;
    var oneof = (_a = update.updateOneof) !== null && _a !== void 0 ? _a : {};
    return {
        filters: (_b = update.filters) !== null && _b !== void 0 ? _b : [],
        createdAt: update.createdAt,
        account: oneof.account,
        slot: oneof.slot,
        transaction: oneof.transaction,
        transactionStatus: oneof.transactionStatus,
        block: oneof.block,
        ping: oneof.ping,
        pong: oneof.pong,
        blockMeta: oneof.blockMeta,
        entry: oneof.entry,
    };
}
function fromJsSubscribeUpdateDeshred(update) {
    var _a, _b;
    var oneof = (_a = update.updateOneof) !== null && _a !== void 0 ? _a : {};
    return {
        filters: (_b = update.filters) !== null && _b !== void 0 ? _b : [],
        createdAt: update.createdAt,
        deshredTransaction: oneof.deshredTransaction,
        ping: oneof.ping,
        pong: oneof.pong,
    };
}
function parseOptionalBool(value, fieldName) {
    if (value === undefined || value === null) {
        return undefined;
    }
    if (typeof value === "boolean") {
        return value;
    }
    if (typeof value === "string") {
        var normalized = value.trim().toLowerCase();
        if (normalized === "true") {
            return true;
        }
        if (normalized === "false") {
            return false;
        }
    }
    throw new Error("Invalid ".concat(fieldName, ": expected true/false, got ").concat(JSON.stringify(value)));
}
function normalizeSubscribeDeshredRequest(request) {
    var _a;
    var normalizedEntries = Object.fromEntries(Object.entries((_a = request.deshredTransactions) !== null && _a !== void 0 ? _a : {}).map(function (_a) {
        var name = _a[0], filter = _a[1];
        return [
            name,
            __assign(__assign({}, filter), { vote: parseOptionalBool(filter.vote, "deshredTransactions.".concat(name, ".vote")) }),
        ];
    }));
    return __assign(__assign({}, request), { deshredTransactions: normalizedEntries });
}
var Client = (function () {
    function Client(endpoint, xToken, channel_options) {
        this._grpcClient = null;
        this._insecureEndpoint = endpoint;
        this._insecureXToken = xToken;
        this._channelOptions = channel_options;
    }
    Client.prototype._connectedGrpcClient = function () {
        if (!this._grpcClient) {
            throw new Error("Client not connected. Call connect() first");
        }
        return this._grpcClient;
    };
    Client.prototype.connect = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this;
                        return [4, napi.GrpcClient.new(this._insecureEndpoint, this._insecureXToken, this._channelOptions)];
                    case 1:
                        _a._grpcClient = _b.sent();
                        return [2];
                }
            });
        });
    };
    Client.prototype.getLatestBlockhash = function (commitment) {
        return __awaiter(this, void 0, void 0, function () {
            var grpcClient, request, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        grpcClient = this._connectedGrpcClient();
                        request = {
                            commitment: commitment !== null && commitment !== void 0 ? commitment : null,
                        };
                        return [4, grpcClient.getLatestBlockhash(request)];
                    case 1:
                        response = _a.sent();
                        return [2, geyser_1.GetLatestBlockhashResponse.fromPartial({
                                slot: response.slot,
                                blockhash: response.blockhash,
                                lastValidBlockHeight: response.lastValidBlockHeight,
                            })];
                }
            });
        });
    };
    Client.prototype.ping = function (count) {
        return __awaiter(this, void 0, void 0, function () {
            var grpcClient, request, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        grpcClient = this._connectedGrpcClient();
                        request = {
                            count: count,
                        };
                        return [4, grpcClient.ping(request)];
                    case 1:
                        response = _a.sent();
                        return [2, geyser_1.PongResponse.fromPartial({
                                count: response.count,
                            })];
                }
            });
        });
    };
    Client.prototype.getBlockHeight = function (commitment) {
        return __awaiter(this, void 0, void 0, function () {
            var grpcClient, request, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        grpcClient = this._connectedGrpcClient();
                        request = {
                            commitment: commitment,
                        };
                        return [4, grpcClient.getBlockHeight(request)];
                    case 1:
                        response = _a.sent();
                        return [2, geyser_1.GetBlockHeightResponse.fromPartial({
                                blockHeight: response.blockHeight,
                            })];
                }
            });
        });
    };
    Client.prototype.getSlot = function (commitment) {
        return __awaiter(this, void 0, void 0, function () {
            var grpcClient, request, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        grpcClient = this._connectedGrpcClient();
                        request = {
                            commitment: commitment,
                        };
                        return [4, grpcClient.getSlot(request)];
                    case 1:
                        response = _a.sent();
                        return [2, geyser_1.GetSlotResponse.fromPartial({
                                slot: response.slot,
                            })];
                }
            });
        });
    };
    Client.prototype.isBlockhashValid = function (blockhash, commitment) {
        return __awaiter(this, void 0, void 0, function () {
            var grpcClient, request, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        grpcClient = this._connectedGrpcClient();
                        request = {
                            blockhash: blockhash,
                            commitment: commitment,
                        };
                        return [4, grpcClient.isBlockhashValid(request)];
                    case 1:
                        response = _a.sent();
                        return [2, geyser_1.IsBlockhashValidResponse.fromPartial({
                                slot: response.slot,
                                valid: response.valid,
                            })];
                }
            });
        });
    };
    Client.prototype.getVersion = function () {
        return __awaiter(this, void 0, void 0, function () {
            var grpcClient, request, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        grpcClient = this._connectedGrpcClient();
                        request = {};
                        return [4, grpcClient.getVersion(request)];
                    case 1:
                        response = _a.sent();
                        return [2, geyser_1.GetVersionResponse.fromPartial({
                                version: response.version,
                            })];
                }
            });
        });
    };
    Client.prototype.subscribeReplayInfo = function () {
        return __awaiter(this, void 0, void 0, function () {
            var grpcClient, request, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        grpcClient = this._connectedGrpcClient();
                        request = {};
                        return [4, grpcClient.subscribeReplayInfo(request)];
                    case 1:
                        response = _a.sent();
                        return [2, geyser_1.SubscribeReplayInfoResponse.fromPartial({
                                firstAvailable: response.firstAvailable,
                            })];
                }
            });
        });
    };
    Client.prototype.subscribe = function () {
        return __awaiter(this, void 0, void 0, function () {
            var grpcClient, options, stream;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        grpcClient = this._connectedGrpcClient();
                        options = {
                            objectMode: true,
                            decodeStrings: false,
                        };
                        return [4, grpcClient.subscribe()];
                    case 1:
                        stream = _a.sent();
                        return [2, new Promise(function (resolve, reject) {
                                try {
                                    resolve(new ClientDuplexStream(stream, options));
                                }
                                catch (err) {
                                    reject(err);
                                }
                            })];
                }
            });
        });
    };
    Client.prototype.subscribeDeshred = function () {
        return __awaiter(this, void 0, void 0, function () {
            var grpcClient, options, stream;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        grpcClient = this._connectedGrpcClient();
                        options = {
                            objectMode: true,
                            decodeStrings: false,
                        };
                        return [4, grpcClient.subscribeDeshred()];
                    case 1:
                        stream = _a.sent();
                        return [2, new Promise(function (resolve, reject) {
                                try {
                                    resolve(new ClientDeshredDuplexStream(stream, options));
                                }
                                catch (err) {
                                    reject(err);
                                }
                            })];
                }
            });
        });
    };
    return Client;
}());
exports.default = Client;
var ClientDuplexStream = (function (_super) {
    __extends(ClientDuplexStream, _super);
    function ClientDuplexStream(stream, options) {
        var _this = _super.call(this, __assign({}, options)) || this;
        _this._napiDuplexStream = stream;
        _this._readInFlight = false;
        _this._isClosed = false;
        _this._isDestroying = false;
        _this._terminalErrorSeen = false;
        _this.once("close", function () {
            _this._isClosed = true;
        });
        return _this;
    }
    ClientDuplexStream.prototype._pullNextUpdate = function () {
        var _this = this;
        if (this._isClosed || this._isDestroying || this._readInFlight) {
            return;
        }
        this._readInFlight = true;
        this._napiDuplexStream
            .read()
            .then(function (update) {
            _this._readInFlight = false;
            if (_this._isClosed || _this._isDestroying) {
                return;
            }
            if (update == null) {
                _this.push(null);
                _this.destroy();
                return;
            }
            var grpcUpdate = fromJsSubscribeUpdate(update);
            var canContinue = _this.push(grpcUpdate);
            if (canContinue) {
                _this._pullNextUpdate();
            }
        })
            .catch(function (err) {
            _this._readInFlight = false;
            if (_this._isClosed || _this._isDestroying) {
                return;
            }
            if (_this._terminalErrorSeen) {
                return;
            }
            _this._terminalErrorSeen = true;
            _this.push(null);
            var terminalError;
            if (err instanceof Error) {
                terminalError = err;
            }
            else if (err &&
                typeof err === "object" &&
                typeof err.message === "string") {
                terminalError = new Error(err.message);
            }
            else {
                terminalError = new Error(String(err !== null && err !== void 0 ? err : "native stream read failed"));
            }
            _this.destroy(terminalError);
        });
    };
    ClientDuplexStream.prototype._read = function (_size) {
        this._pullNextUpdate();
    };
    ClientDuplexStream.prototype._destroy = function (error, callback) {
        this._isDestroying = true;
        this._isClosed = true;
        this._terminalErrorSeen = true;
        try {
            var nativeStream = this._napiDuplexStream;
            if (typeof nativeStream.close === "function") {
                nativeStream.close();
            }
        }
        catch (_a) { }
        callback(error);
    };
    ClientDuplexStream.prototype._write = function (chunk, _encoding, callback) {
        if (this._isClosed || this._isDestroying) {
            callback(new Error("Cannot write to a closed subscription stream"));
            return;
        }
        try {
            var encodedRequest = geyser_1.SubscribeRequest.encode(chunk).finish();
            var nativeStream = this._napiDuplexStream;
            if (typeof nativeStream.writeRaw !== "function") {
                throw new Error("Native stream does not support writeRaw; reinstall the SDK so JS and native bindings match");
            }
            nativeStream.writeRaw(encodedRequest);
            callback();
        }
        catch (err) {
            callback(err);
        }
    };
    return ClientDuplexStream;
}(stream_1.Duplex));
exports.ClientDuplexStream = ClientDuplexStream;
var ClientDeshredDuplexStream = (function (_super) {
    __extends(ClientDeshredDuplexStream, _super);
    function ClientDeshredDuplexStream(stream, options) {
        var _this = _super.call(this, __assign({}, options)) || this;
        _this._napiDuplexStream = stream;
        _this._readInFlight = false;
        _this._isClosed = false;
        _this._isDestroying = false;
        _this._terminalErrorSeen = false;
        _this.once("close", function () {
            _this._isClosed = true;
        });
        return _this;
    }
    ClientDeshredDuplexStream.prototype._pullNextUpdate = function () {
        var _this = this;
        if (this._isClosed || this._isDestroying || this._readInFlight) {
            return;
        }
        this._readInFlight = true;
        this._napiDuplexStream
            .read()
            .then(function (update) {
            _this._readInFlight = false;
            if (_this._isClosed || _this._isDestroying) {
                return;
            }
            if (update == null) {
                _this.push(null);
                _this.destroy();
                return;
            }
            var grpcUpdate = fromJsSubscribeUpdateDeshred(update);
            var canContinue = _this.push(grpcUpdate);
            if (canContinue) {
                _this._pullNextUpdate();
            }
        })
            .catch(function (err) {
            _this._readInFlight = false;
            if (_this._isClosed || _this._isDestroying) {
                return;
            }
            if (_this._terminalErrorSeen) {
                return;
            }
            _this._terminalErrorSeen = true;
            _this.push(null);
            var terminalError;
            if (err instanceof Error) {
                terminalError = err;
            }
            else if (err &&
                typeof err === "object" &&
                typeof err.message === "string") {
                terminalError = new Error(err.message);
            }
            else {
                terminalError = new Error(String(err !== null && err !== void 0 ? err : "native stream read failed"));
            }
            _this.destroy(terminalError);
        });
    };
    ClientDeshredDuplexStream.prototype._read = function (_size) {
        this._pullNextUpdate();
    };
    ClientDeshredDuplexStream.prototype._destroy = function (error, callback) {
        this._isDestroying = true;
        this._isClosed = true;
        this._terminalErrorSeen = true;
        try {
            var nativeStream = this._napiDuplexStream;
            if (typeof nativeStream.close === "function") {
                nativeStream.close();
            }
        }
        catch (_a) { }
        callback(error);
    };
    ClientDeshredDuplexStream.prototype._write = function (chunk, _encoding, callback) {
        if (this._isClosed || this._isDestroying) {
            callback(new Error("Cannot write to a closed deshred subscription stream"));
            return;
        }
        try {
            var normalizedChunk = normalizeSubscribeDeshredRequest(chunk);
            var encodedRequest = geyser_1.SubscribeDeshredRequest.encode(normalizedChunk).finish();
            var nativeStream = this._napiDuplexStream;
            if (typeof nativeStream.writeRaw !== "function") {
                throw new Error("Native stream does not support writeRaw; reinstall the SDK so JS and native bindings match");
            }
            nativeStream.writeRaw(encodedRequest);
            callback();
        }
        catch (err) {
            callback(err);
        }
    };
    return ClientDeshredDuplexStream;
}(stream_1.Duplex));
exports.ClientDeshredDuplexStream = ClientDeshredDuplexStream;
exports.txEncode = {
    encoding: napi.WasmUiTransactionEncoding,
    encode_raw: napi.encodeTx,
    encode: function (message, encoding, max_supported_transaction_version, show_rewards) {
        return JSON.parse(napi.encodeTx(geyser_1.SubscribeUpdateTransactionInfo.encode(message).finish(), encoding, max_supported_transaction_version, show_rewards));
    },
};
exports.txDeshredEncode = {
    encoding: napi.WasmUiTransactionEncoding,
    encode_raw: napi.encodeDeshredTx,
    encode: function (message, encoding) {
        return JSON.parse(napi.encodeDeshredTx(geyser_1.SubscribeUpdateDeshredTransactionInfo.encode(message).finish(), encoding));
    },
};
exports.txErrDecode = {
    decode_raw: napi.decodeTxError,
    decode: function (buf) {
        return JSON.parse(napi.decodeTxError(Array.from(buf)));
    },
};
