// @triton-one/yellowstone-grpc/dist/esm/index.js
/** TypeScript/JavaScript client for gRPC Geyser. */
// Import generated gRPC client and types.
import { GetBlockHeightResponse as GetBlockHeightResponseMessage, GetLatestBlockhashResponse as GetLatestBlockhashResponseMessage, GetSlotResponse as GetSlotResponseMessage, GetVersionResponse as GetVersionResponseMessage, IsBlockhashValidResponse as IsBlockhashValidResponseMessage, PongResponse as PongResponseMessage, SubscribeDeshredRequest as SubscribeDeshredRequestMessage, SubscribeRequest as SubscribeRequestMessage, SubscribeReplayInfoResponse as SubscribeReplayInfoResponseMessage, SubscribeUpdateDeshredTransactionInfo, SubscribeUpdateTransactionInfo, } from "./grpc/geyser.js";
// Reexport automatically generated types
export { CommitmentLevel, SubscribeDeshredRequest, SubscribeDeshredRequest_DeshredTransactionsEntry, SubscribeRequest, SubscribeRequest_AccountsEntry, SubscribeRequest_BlocksEntry, SubscribeRequest_BlocksMetaEntry, SubscribeRequestFilterDeshredTransactions, SubscribeRequest_SlotsEntry, SubscribeRequest_TransactionsEntry, SubscribeRequestAccountsDataSlice, SubscribeRequestFilterAccounts, SubscribeRequestFilterAccountsFilter, SubscribeRequestFilterAccountsFilterLamports, SubscribeRequestFilterAccountsFilterMemcmp, SubscribeRequestFilterBlocks, SubscribeRequestFilterBlocksMeta, SubscribeRequestFilterEntry, SubscribeRequestFilterSlots, SubscribeRequestFilterTransactions, SubscribeUpdate, SubscribeUpdateAccount, SubscribeUpdateAccountInfo, SubscribeUpdateBlock, SubscribeUpdateBlockMeta, SubscribeUpdateDeshred, SubscribeUpdateDeshredTransaction, SubscribeUpdateDeshredTransactionInfo, SubscribeUpdatePing, SubscribeUpdateSlot, SubscribeUpdateTransaction, SubscribeUpdateTransactionInfo, } from "./grpc/geyser.js";
import { Duplex } from "stream";
import * as napi from "./napi/index.js";
/**
 * Convert N-API `JsSubscribeUpdate` shape (with `updateOneof`) into the
 * generated protobuf-friendly SDK shape (`SubscribeUpdate`) where the oneof
 * variants are top-level optional fields.
 */
function fromJsSubscribeUpdate(update) {
    const oneof = update.updateOneof ?? {};
    return {
        filters: update.filters ?? [],
        createdAt: update.createdAt,
        account: oneof.account,
        slot: oneof.slot,
        transaction: oneof.transaction,
        transactionStatus: oneof.transactionStatus,
        block: oneof.block,
        ping: oneof.ping,
        pong: oneof.pong,
        blockMeta: oneof.blockMeta,
        entry: oneof.entry,
    };
}
/**
 * Convert N-API `JsSubscribeUpdateDeshred` shape (with `updateOneof`) into the
 * generated protobuf-friendly SDK shape (`SubscribeUpdateDeshred`).
 */
function fromJsSubscribeUpdateDeshred(update) {
    const oneof = update.updateOneof ?? {};
    return {
        filters: update.filters ?? [],
        createdAt: update.createdAt,
        deshredTransaction: oneof.deshredTransaction,
        ping: oneof.ping,
        pong: oneof.pong,
    };
}
function parseOptionalBool(value, fieldName) {
    if (value === undefined || value === null) {
        return undefined;
    }
    if (typeof value === "boolean") {
        return value;
    }
    if (typeof value === "string") {
        const normalized = value.trim().toLowerCase();
        if (normalized === "true") {
            return true;
        }
        if (normalized === "false") {
            return false;
        }
    }
    throw new Error(`Invalid ${fieldName}: expected true/false, got ${JSON.stringify(value)}`);
}
function normalizeSubscribeDeshredRequest(request) {
    const normalizedEntries = Object.fromEntries(Object.entries(request.deshredTransactions ?? {}).map(([name, filter]) => [
        name,
        {
            ...filter,
            vote: parseOptionalBool(filter.vote, `deshredTransactions.${name}.vote`),
        },
    ]));
    return {
        ...request,
        deshredTransactions: normalizedEntries,
    };
}
export default class Client {
    _insecureEndpoint;
    _insecureXToken;
    _channelOptions;
    _grpcClient = null;
    constructor(endpoint, xToken, channel_options) {
        this._insecureEndpoint = endpoint;
        this._insecureXToken = xToken;
        this._channelOptions = channel_options;
    }
    _connectedGrpcClient() {
        if (!this._grpcClient) {
            throw new Error("Client not connected. Call connect() first");
        }
        return this._grpcClient;
    }
    async connect() {
        // Establish one persistent native gRPC client reused by all calls.
        this._grpcClient = await napi.GrpcClient.new(this._insecureEndpoint, this._insecureXToken, this._channelOptions);
    }
    async getLatestBlockhash(commitment) {
        const grpcClient = this._connectedGrpcClient();
        const request = {
            commitment: commitment ?? null,
        };
        const response = await grpcClient.getLatestBlockhash(request);
        return GetLatestBlockhashResponseMessage.fromPartial({
            slot: response.slot,
            blockhash: response.blockhash,
            lastValidBlockHeight: response.lastValidBlockHeight,
        });
    }
    async ping(count) {
        const grpcClient = this._connectedGrpcClient();
        const request = {
            count,
        };
        const response = await grpcClient.ping(request);
        return PongResponseMessage.fromPartial({
            count: response.count,
        });
    }
    async getBlockHeight(commitment) {
        const grpcClient = this._connectedGrpcClient();
        const request = {
            commitment,
        };
        const response = await grpcClient.getBlockHeight(request);
        return GetBlockHeightResponseMessage.fromPartial({
            blockHeight: response.blockHeight,
        });
    }
    async getSlot(commitment) {
        const grpcClient = this._connectedGrpcClient();
        const request = {
            commitment,
        };
        const response = await grpcClient.getSlot(request);
        return GetSlotResponseMessage.fromPartial({
            slot: response.slot,
        });
    }
    async isBlockhashValid(blockhash, commitment) {
        const grpcClient = this._connectedGrpcClient();
        const request = {
            blockhash,
            commitment,
        };
        const response = await grpcClient.isBlockhashValid(request);
        return IsBlockhashValidResponseMessage.fromPartial({
            slot: response.slot,
            valid: response.valid,
        });
    }
    async getVersion() {
        const grpcClient = this._connectedGrpcClient();
        const request = {};
        const response = await grpcClient.getVersion(request);
        return GetVersionResponseMessage.fromPartial({
            version: response.version,
        });
    }
    async subscribeReplayInfo() {
        const grpcClient = this._connectedGrpcClient();
        const request = {};
        const response = await grpcClient.subscribeReplayInfo(request);
        return SubscribeReplayInfoResponseMessage.fromPartial({
            firstAvailable: response.firstAvailable,
        });
    }
    async subscribe() {
        const grpcClient = this._connectedGrpcClient();
        // Inner stream.Duplex config passed to both stream.Readable and Writable.
        // See: https://nodejs.org/en/blog/feature/streams2#new-streamduplexoptions
        const options = {
            // Pass objects not bytes.
            objectMode: true,
            // Skip unnecessary buffer conversion for performance.
            decodeStrings: false,
            // TODO: Fine tune high watermark for performance and backpressure.
            // highWaterMark: 16
        };
        // Native stream produces N-API generated JS objects; wrapper below adapts
        // to public protobuf-generated SDK shapes and Node stream semantics.
        const stream = await grpcClient.subscribe();
        return new Promise((resolve, reject) => {
            try {
                resolve(new ClientDuplexStream(stream, options));
            }
            catch (err) {
                reject(err);
            }
        });
    }
    async subscribeDeshred() {
        const grpcClient = this._connectedGrpcClient();
        // Same stream options used by regular subscribe() wrapper.
        const options = {
            objectMode: true,
            decodeStrings: false,
        };
        // Native layer opens the gRPC stream before resolving this Promise, so
        // unsupported RPCs (UNIMPLEMENTED) throw here and bubble to caller.
        const stream = await grpcClient.subscribeDeshred();
        return new Promise((resolve, reject) => {
            try {
                resolve(new ClientDeshredDuplexStream(stream, options));
            }
            catch (err) {
                reject(err);
            }
        });
    }
}
export class ClientDuplexStream extends Duplex {
    _napiDuplexStream;
    // Prevent overlapping native reads: a single pending read at a time.
    _readInFlight;
    // Closed once Node emits `close`.
    _isClosed;
    // Set during destroy path to short-circuit late async completions.
    _isDestroying;
    // Ensure we surface at most one terminal error per stream instance.
    _terminalErrorSeen;
    constructor(stream, options) {
        super({ ...options });
        this._napiDuplexStream = stream;
        this._readInFlight = false;
        this._isClosed = false;
        this._isDestroying = false;
        this._terminalErrorSeen = false;
        this.once("close", () => {
            this._isClosed = true;
        });
    }
    _pullNextUpdate() {
        if (this._isClosed || this._isDestroying || this._readInFlight) {
            return;
        }
        this._readInFlight = true;
        this._napiDuplexStream
            .read()
            .then((update) => {
            this._readInFlight = false;
            if (this._isClosed || this._isDestroying) {
                return;
            }
            // Native side can signal EOF by returning no update.
            if (update == null) {
                this.push(null);
                this.destroy();
                return;
            }
            const grpcUpdate = fromJsSubscribeUpdate(update);
            // Respect backpressure: only pull again if consumer accepted push.
            const canContinue = this.push(grpcUpdate);
            if (canContinue) {
                this._pullNextUpdate();
            }
        })
            .catch((err) => {
            this._readInFlight = false;
            if (this._isClosed || this._isDestroying) {
                return;
            }
            if (this._terminalErrorSeen) {
                return;
            }
            this._terminalErrorSeen = true;
            this.push(null); // Signal end of stream
            let terminalError;
            if (err instanceof Error) {
                terminalError = err;
            }
            else if (err &&
                typeof err === "object" &&
                typeof err.message === "string") {
                terminalError = new Error(err.message);
            }
            else {
                terminalError = new Error(String(err ?? "native stream read failed"));
            }
            this.destroy(terminalError);
        });
    }
    _read(_size) {
        this._pullNextUpdate();
    }
    _destroy(error, callback) {
        // Mark terminal state first so late read completions are ignored.
        this._isDestroying = true;
        this._isClosed = true;
        this._terminalErrorSeen = true;
        // Explicitly stop the native worker so it does not outlive JS stream state.
        try {
            const nativeStream = this._napiDuplexStream;
            if (typeof nativeStream.close === "function") {
                nativeStream.close();
            }
        }
        catch { }
        callback(error);
    }
    _write(chunk, _encoding, callback) {
        if (this._isClosed || this._isDestroying) {
            callback(new Error("Cannot write to a closed subscription stream"));
            return;
        }
        try {
            const encodedRequest = SubscribeRequestMessage.encode(chunk).finish();
            const nativeStream = this._napiDuplexStream;
            if (typeof nativeStream.writeRaw !== "function") {
                throw new Error("Native stream does not support writeRaw; reinstall the SDK so JS and native bindings match");
            }
            nativeStream.writeRaw(encodedRequest);
            callback();
        }
        catch (err) {
            callback(err);
        }
    }
}
export class ClientDeshredDuplexStream extends Duplex {
    _napiDuplexStream;
    _readInFlight;
    _isClosed;
    _isDestroying;
    _terminalErrorSeen;
    constructor(stream, options) {
        super({ ...options });
        this._napiDuplexStream = stream;
        this._readInFlight = false;
        this._isClosed = false;
        this._isDestroying = false;
        this._terminalErrorSeen = false;
        this.once("close", () => {
            this._isClosed = true;
        });
    }
    _pullNextUpdate() {
        if (this._isClosed || this._isDestroying || this._readInFlight) {
            return;
        }
        this._readInFlight = true;
        this._napiDuplexStream
            .read()
            .then((update) => {
            this._readInFlight = false;
            if (this._isClosed || this._isDestroying) {
                return;
            }
            if (update == null) {
                this.push(null);
                this.destroy();
                return;
            }
            const grpcUpdate = fromJsSubscribeUpdateDeshred(update);
            const canContinue = this.push(grpcUpdate);
            if (canContinue) {
                this._pullNextUpdate();
            }
        })
            .catch((err) => {
            this._readInFlight = false;
            if (this._isClosed || this._isDestroying) {
                return;
            }
            if (this._terminalErrorSeen) {
                return;
            }
            this._terminalErrorSeen = true;
            this.push(null);
            let terminalError;
            if (err instanceof Error) {
                terminalError = err;
            }
            else if (err &&
                typeof err === "object" &&
                typeof err.message === "string") {
                terminalError = new Error(err.message);
            }
            else {
                terminalError = new Error(String(err ?? "native stream read failed"));
            }
            this.destroy(terminalError);
        });
    }
    _read(_size) {
        this._pullNextUpdate();
    }
    _destroy(error, callback) {
        this._isDestroying = true;
        this._isClosed = true;
        this._terminalErrorSeen = true;
        try {
            const nativeStream = this._napiDuplexStream;
            if (typeof nativeStream.close === "function") {
                nativeStream.close();
            }
        }
        catch { }
        callback(error);
    }
    _write(chunk, _encoding, callback) {
        if (this._isClosed || this._isDestroying) {
            callback(new Error("Cannot write to a closed deshred subscription stream"));
            return;
        }
        try {
            const normalizedChunk = normalizeSubscribeDeshredRequest(chunk);
            const encodedRequest = SubscribeDeshredRequestMessage.encode(normalizedChunk).finish();
            const nativeStream = this._napiDuplexStream;
            if (typeof nativeStream.writeRaw !== "function") {
                throw new Error("Native stream does not support writeRaw; reinstall the SDK so JS and native bindings match");
            }
            nativeStream.writeRaw(encodedRequest);
            callback();
        }
        catch (err) {
            callback(err);
        }
    }
}
export const txEncode = {
    encoding: napi.WasmUiTransactionEncoding,
    encode_raw: napi.encodeTx,
    encode: (message, encoding, max_supported_transaction_version, show_rewards) => {
        return JSON.parse(napi.encodeTx(SubscribeUpdateTransactionInfo.encode(message).finish(), encoding, max_supported_transaction_version, show_rewards));
    },
};
export const txDeshredEncode = {
    encoding: napi.WasmUiTransactionEncoding,
    encode_raw: napi.encodeDeshredTx,
    encode: (message, encoding) => {
        return JSON.parse(napi.encodeDeshredTx(SubscribeUpdateDeshredTransactionInfo.encode(message).finish(), encoding));
    },
};
export const txErrDecode = {
    decode_raw: napi.decodeTxError,
    decode: (buf) => {
        return JSON.parse(napi.decodeTxError(Array.from(buf)));
    },
};
